File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/well_known.py
"""App Link / Universal Link binding documents, built from ``AttestationConfig``.

Config-driven and env-free: the host supplies ``ios_applink_app_id`` /
``ios_app_id`` / ``ios_app_clip_id`` / ``android_package_name`` /
``android_sha256_cert_fingerprints`` / ``applink_path`` via the injected config,
so a fresh deployment only changes settings to bind a new app identity.
"""

from __future__ import annotations

from typing import Any

from .config import get_attestation_config


def applink_path() -> str:
    return get_attestation_config().applink_path or "/native*"


def apple_app_site_association() -> dict[str, Any]:
    """Build the AASA document iOS fetches from /.well-known to bind the domain."""
    cfg = get_attestation_config()
    app_id = (cfg.ios_applink_app_id or cfg.ios_app_id or "").strip()
    clip_id = (cfg.ios_app_clip_id or "").strip()

    app_ids: list[str] = []
    if app_id:
        app_ids.append(app_id)
    if clip_id:
        app_ids.append(clip_id)

    doc: dict[str, Any] = {
        "applinks": {
            "details": [
                {
                    "appIDs": app_ids,
                    "components": [
                        {
                            "/": applink_path(),
                            "comment": "Opens the app for QuickLink liveness sessions.",
                        }
                    ],
                }
            ]
        }
    }
    if clip_id:
        doc["appclips"] = {"apps": [clip_id]}
    return doc


def assetlinks() -> list[dict[str, Any]]:
    """Build the Digital Asset Links document Android fetches to bind the domain."""
    cfg = get_attestation_config()
    package_name = (cfg.android_package_name or "").strip()
    fingerprints = list(cfg.android_sha256_cert_fingerprints)

    return [
        {
            "relation": ["delegate_permission/common.handle_all_urls"],
            "target": {
                "namespace": "android_app",
                "package_name": package_name,
                "sha256_cert_fingerprints": fingerprints,
            },
        }
    ]

File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Services/WellKnown.cs
using System.Text.Json.Nodes;
using Azure.AI.Vision.Face.DeviceAttestation.Configuration;

namespace Azure.AI.Vision.Face.DeviceAttestation.Services;

/// <summary>
/// App Link / Universal Link binding documents served at /.well-known/*, driven
/// entirely by the injected <see cref="AttestationConfig"/> so the library binds
/// to any domain / app identity purely through configuration.
/// </summary>
internal static class WellKnown
{
    /// <summary>Build the AASA document iOS fetches from /.well-known to bind the domain.</summary>
    public static JsonObject AppleAppSiteAssociation(AttestationConfig cfg)
    {
        var appId = (!string.IsNullOrEmpty(cfg.IosApplinkAppId) ? cfg.IosApplinkAppId! : cfg.IosAppId).Trim();
        var clipId = (cfg.IosAppClipId ?? "").Trim();

        var appIds = new JsonArray();
        if (appId.Length > 0)
        {
            appIds.Add(appId);
        }
        if (clipId.Length > 0)
        {
            appIds.Add(clipId);
        }

        var doc = new JsonObject
        {
            ["applinks"] = new JsonObject
            {
                ["details"] = new JsonArray
                {
                    new JsonObject
                    {
                        ["appIDs"] = appIds,
                        ["components"] = new JsonArray
                        {
                            new JsonObject
                            {
                                ["/"] = cfg.ApplinkPath,
                                ["comment"] = "Opens the app for liveness sessions.",
                            },
                        },
                    },
                },
            },
        };

        if (clipId.Length > 0)
        {
            doc["appclips"] = new JsonObject { ["apps"] = new JsonArray { clipId } };
        }
        return doc;
    }

    /// <summary>Build the Digital Asset Links document Android fetches to bind the domain.</summary>
    public static JsonArray AssetLinks(AttestationConfig cfg)
    {
        var packageName = (cfg.AndroidPackageName ?? "").Trim();

        var fingerprints = new JsonArray();
        foreach (var fp in cfg.AndroidSha256CertFingerprints)
        {
            fingerprints.Add(fp);
        }

        return new JsonArray
        {
            new JsonObject
            {
                ["relation"] = new JsonArray { "delegate_permission/common.handle_all_urls" },
                ["target"] = new JsonObject
                {
                    ["namespace"] = "android_app",
                    ["package_name"] = packageName,
                    ["sha256_cert_fingerprints"] = fingerprints,
                },
            },
        };
    }
}
